x/ibc-rate-limit/ibc_middleware.go

```diff
@@ -0,0 +1,251 @@
+package ibc_rate_limit
+
+import (
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v3/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+	"github.com/osmosis-labs/osmosis/v11/x/ibc-rate-limit/types"
+	lockupkeeper "github.com/osmosis-labs/osmosis/v11/x/lockup/keeper"
+)
+
+var (
+	_ porttypes.Middleware  = &IBCModule{}
+	_ porttypes.ICS4Wrapper = &ICS4Middleware{}
+)
+
+type ICS4Middleware struct {
+	channel       porttypes.ICS4Wrapper
+	accountKeeper *authkeeper.AccountKeeper
+	BankKeeper    *bankkeeper.BaseKeeper
+	WasmKeeper    *wasmkeeper.Keeper
+	LockupKeeper  *lockupkeeper.Keeper
+	ParamSpace    paramtypes.Subspace
```

Comment:
I tend to be a proponent of the other camp where we hide as much information as possible.

In my opinion, this helps with encapsulating all internal components and leads to an easier experience with reasoning about an abstraction.

Also, I understand the "we'are all adults" idea if it is in the scope of a limited group of team members working on the project. Since we are in the open source where there are many contributors with varying commitments as well as reviewers with limited capacity, I would prefer to err on the side of caution and avoid exposing interfaces that can be private.

Also, this middleware seems to be a similar abstraction to a `keeper`. As far as I know, most of our keepers aim to maximize information hiding by unexporting fields that are unused in other packages. So it would be great to aim to keep middleware consistent with keepers.

Please let me know what you think

Reply:
yeah, that makes sense. I think if osmosis was a library that others will import (the sdk, wasmd, etc) I'd feel more strongly about allowing implementors to change internals, but for this it makes sense to clearly signal which parts of the internals are stable and which aren't. I'll see how much I can hide.

Reply:
changed.